packages/gatsby-plugin-netlify/src/build-headers-program.js

```diff
@@ -89,15 +97,24 @@ function preloadHeadersByPage(pages, manifest, pathPrefix) {
   let linksByPage = {}
 
   pages.forEach(page => {
-    const scripts = [
-      ...COMMON_BUNDLES,
-      pathChunkName(page.path),
-      page.componentChunkName,
+    const scripts = _.flatMap(COMMON_BUNDLES, file =>
+      getScriptPath(file, manifest)
+    )
+    scripts.push(...getScriptPath(pathChunkName(page.path), manifest))
+    scripts.push(...getScriptPath(page.componentChunkName, manifest))
+
+    const json = [
+      posix.join(PAGE_DATA_DIR, `app-data.json`),
+      getPageDataPath(page.path),
```

Comment:
this assumes that core version is at least 2.17.0 (that's when app-data was introduced)

what would happen if app-data (or even page-data, which was added in 2.9.0) is not actually there?

Reply:
good call, we get a 404, I'll do a file exists check so we're sure and measure how long it would take. A gatsby version check would make things faster but unsure if that's what we want.

Reply:
yeah, staying on v2 for so long makes it super hard to keep maintaining compatibility for (much) older gatsby versions in plugins ;( I don't have good answer here, maybe this is most OK to keep it as is - of course making requests that end up in 404 is not great scenario, but then again, neither are `fs.exists` checks, and version check while would be the fastest - it's also very brittle

What about this:
on first page we handle, we do `fs.exists` checks on both page-data and app-data and set flag saying `gatsbyUsesPageData`/`gatsbyUsesAppData` and we use that on subsequent calls to not have excessive fs checks for page-data for all files?

Reply:
oh yeah that sounds good great, it's like a feature exists check! 👏